arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java

```diff
@@ -0,0 +1,387 @@
+  public static final int UNKNOWN_WIDTH = -1;
+
+  private final ColumnDescriptor columnDescriptor;
+  private final int batchSize;
+  private final VectorizedColumnIterator vectorizedColumnIterator;
+  private final boolean isFixedLengthDecimal;
+  private final boolean isVarWidthType;
+  private final boolean isFixedWidthBinary;
+  private final boolean isBooleanType;
+  private final boolean isPaddedDecimal;
+  private final boolean isIntType;
+  private final boolean isLongType;
+  private final boolean isFloatType;
+  private final boolean isDoubleType;
+  private final Types.NestedField icebergField;
+  private final BufferAllocator rootAlloc;
+  private FieldVector vec;
+  private int typeWidth;
+  private boolean reuseContainers = true;
+  private NullabilityHolder nullabilityHolder;
+
+  // In cases when Parquet employs fall back to plain encoding, we eagerly decode the dictionary encoded pages
+  // before storing the values in the Arrow vector. This means even if the dictionary is present, data
+  // present in the vector may not necessarily be dictionary encoded.
+  private Dictionary dictionary;
+  private boolean allPagesDictEncoded;
+
+  // This value is copied from Arrow's BaseVariableWidthVector. We may need to change
+  // this value if Arrow ends up changing this default.
+  private static final int DEFAULT_RECORD_BYTE_COUNT = 8;
+
+  public VectorizedArrowReader(
+      ColumnDescriptor desc,
+      Types.NestedField icebergField,
+      BufferAllocator ra,
+      int batchSize) {
+    this.icebergField = icebergField;
+    this.batchSize = (batchSize == 0) ? DEFAULT_BATCH_SIZE : batchSize;
+    this.columnDescriptor = desc;
+    this.rootAlloc = ra;
+    this.isFixedLengthDecimal = isFixedLengthDecimal(desc);
+    this.isVarWidthType = isVarWidthType(desc);
+    this.isFixedWidthBinary = isFixedWidthBinary(desc);
+    this.isBooleanType = isBooleanType(desc);
+    this.isPaddedDecimal = isIntLongBackedDecimal(desc);
+    this.isIntType = isIntType(desc);
+    this.isLongType = isLongType(desc);
+    this.isFloatType = isFloatType(desc);
+    this.isDoubleType = isDoubleType(desc);
+    this.vectorizedColumnIterator = new VectorizedColumnIterator(desc, "", batchSize);
+  }
+
+  private VectorizedArrowReader() {
+    this.icebergField = null;
+    this.batchSize = DEFAULT_BATCH_SIZE;
+    this.columnDescriptor = null;
+    this.rootAlloc = null;
+    this.isFixedLengthDecimal = false;
+    this.isVarWidthType = false;
+    this.isFixedWidthBinary = false;
+    this.isBooleanType = false;
+    this.isPaddedDecimal = false;
+    this.isIntType = false;
+    this.isLongType = false;
+    this.isFloatType = false;
+    this.isDoubleType = false;
+    this.vectorizedColumnIterator = null;
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  @Override
+  public VectorHolder read(int numValsToRead) {
+    if (vec == null || !reuseContainers) {
+      typeWidth = allocateFieldVector();
```

Comment:
Right now, allocation and reads use separate logic. Reads use the `isXType` booleans, while `allocateFieldVector` uses the Parquet type.

I think I understand why those are slightly different: allocation needs to create, for example, a `DateDayVector` but reading calls `nextBatchIntergers` because it writes into the underlying buffer directly.

I think that having these completely decoupled could lead to problems, though. I'd rather allocate a vector and then base the type of read on the vector that was produced. Something like this:

```
  public VectorizedArrowReader(...) {
    ...
    this.vector = allocateFieldVector();
    if (this.vector instanceof IntVector || this.vector instanceof DateDayVector) {
      this.readMethod = ReadMethod.INT;
    } else if (this.vector instanceof LongVector || this.vector instanceof TimeStampMicroTZVector) {
      this.readMethod = ReadMethod.LONG;
    }
  }
```

The logic would probably be more complicated because it needs to rely on both the incoming Parquet encoding and the outgoing vector type. But the main thing is that it would be good to check that the vector and the read method are compatible.